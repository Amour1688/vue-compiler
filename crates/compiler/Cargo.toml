[package]
name = "compiler"
version = "0.1.0"
authors = ["Herrington Darkholme <2883231+HerringtonDarkholme@users.noreply.github.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
smallvec = "1.6.1"
bitflags = "1.3"
rustc-hash = "1.1.0"
json = "0.12.4"
phf = { version = "0.10", features = ["macros"] }

[dev-dependencies]
criterion = "0.3"
insta = "1.8.0"
serde = { version = "1.0", features = ["derive"] }
smallvec = {version="1.2.0", features = ["serde"]}
