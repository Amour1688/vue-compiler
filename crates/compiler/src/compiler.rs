use super::{
    codegen::{
        CodeGenerateOption, CodeGenerator, CodeWriter, DecodedStr, EntityDecoder, ScriptMode,
    },
    converter::{
        no_op_directive_convert, BaseConvertInfo as BaseInfo, BaseConverter, BaseRoot,
        ConvertOption, Converter, DirConvertFn, V_BIND, V_MODEL,
    },
    error::{NoopErrorHandler, RcErrHandle},
    flags::RuntimeHelper,
    parser::{Element, ParseOption, Parser, WhitespaceStrategy, AstRoot},
    scanner::{ScanOption, Scanner, TextMode, Tokens},
    transformer::{BaseTransformer, CorePass, TransformOption, Transformer},
    util::{no, yes},
    Namespace,
};

use rustc_hash::FxHashMap;
use std::{io, rc::Rc};

pub struct CompileOption {
    /// e.g. platform native elements, e.g. `<div>` for browsers
    pub is_native_tag: fn(&str) -> bool,

    /// e.g. native elements that can self-close, e.g. `<img>`, `<br>`, `<hr>`
    pub is_void_tag: fn(&str) -> bool,

    /// e.g. elements that should preserve whitespace inside, e.g. `<pre>`
    pub is_pre_tag: fn(&str) -> bool,

    /// Platform-specific built-in components e.g. `<Transition>`
    /// The pairing runtime provides additional built-in elements,
    /// Platform developer can use this to mark them as built-in
    /// so the compiler will generate component vnodes for them.
    pub get_builtin_component: fn(&str) -> Option<RuntimeHelper>,

    /// Separate option for end users to extend the native elements list
    pub is_custom_element: fn(&str) -> bool,

    /// Get tag namespace
    pub get_namespace: fn(&str, &Vec<Element<'_>>) -> Namespace,

    /// Get text parsing mode for this element
    pub get_text_mode: fn(&str) -> TextMode,

    /// @default ['{{', '}}']
    pub delimiters: (String, String),

    /// Whitespace handling strategy
    pub whitespace: WhitespaceStrategy,

    /// Only needed for DOM compilers
    pub decode_entities: EntityDecoder,

    /// Whether to keep comments in the templates AST.
    /// This defaults to `true` in development and `false` in production builds.
    pub preserve_comments: Option<bool>,
    /// Whether the output is dev build which includes v-if comment and dev patch flags.
    pub is_dev: bool,

    /// An object of { name: transform } to be applied to every directive attribute
    /// node found on element nodes.
    pub directive_converters: FxHashMap<&'static str, DirConvertFn>,
    /// Hoist static VNodes and props objects to `_hoisted_x` constants
    /// @default false
    pub hoist_static: bool,
    /// Cache v-on handlers to avoid creating new inline functions on each render,
    /// also avoids the need for dynamically patching the handlers by wrapping it.
    /// e.g `@click="foo"` by default is compiled to `{ onClick: foo }`. With this
    /// option it's compiled to:
    /// ```js
    /// { onClick: _cache[0] || (_cache[0] = e => _ctx.foo(e)) }
    /// ```
    /// - Requires "prefixIdentifiers" to be enabled because it relies on scope
    /// analysis to determine if a handler is safe to cache.
    /// @default false
    pub cache_handlers: bool,

    /// - `module` mode will generate ES module import statements for helpers
    /// and export the render function as the default export.
    /// - `function` mode will generate a single `const { helpers... } = Vue`
    /// statement and return the render function. It expects `Vue` to be globally
    /// available (or passed by wrapping the code with an IIFE). It is meant to be
    /// used with `new Function(code)()` to generate a render function at runtime.
    /// @default 'function'
    pub mode: ScriptMode,
    /// Generate source map?
    /// @default false
    pub source_map: bool,
    /// Whether the output JS needs re-rendering when Vue runtime data change.
    /// e.g. SSR can set it to false since SSR is executed only once per request.
    /// @default true
    pub need_reactivity: bool,
    /// Custom error reporter. Default is noop.
    pub error_handler: RcErrHandle,
    // deleted options
    // nodeTransforms?: NodeTransform[]
    // transformHoist?: HoistTransform | null
    // expressionPlugins?: ParserPlugin[]
    // prefix_identifiers: bool,
    // optimizeImports?: boolean // farewell, webpack optimization

    // moved to SFCInfo
    // bindingMetadata?: BindingMetadata
    // inline?: boolean
    // filename?: string
    // scopeId?: string | null
    // slotted?: boolean

    // moved to SSR or need_reactivity
    // ssr: bool // will be false in fallback node
    // inSSR?: bool // always true in ssr build
    // ssrCssVars?: string
    // ssrRuntimeModuleName?: string
}

impl Default for CompileOption {
    fn default() -> Self {
        let mut directive_converters = FxHashMap::default();
        directive_converters.insert(V_BIND.0, V_BIND.1);
        directive_converters.insert(V_MODEL.0, V_MODEL.1);
        directive_converters.insert("on", no_op_directive_convert);
        Self {
            is_native_tag: yes,
            is_void_tag: no,
            is_pre_tag: no,
            get_builtin_component: |_| None,
            is_custom_element: no,
            get_namespace: |_, _| Namespace::Html,
            get_text_mode: |_| TextMode::Data,
            delimiters: ("{{".into(), "}}".into()),
            whitespace: WhitespaceStrategy::Preserve,
            decode_entities: |s, _| DecodedStr::from(s),
            preserve_comments: None,
            is_dev: true,
            directive_converters,
            hoist_static: false,
            cache_handlers: false,
            mode: ScriptMode::Function {
                prefix_identifier: false,
                runtime_global_name: "Vue".into(),
            },
            source_map: false,
            need_reactivity: true,
            error_handler: Rc::new(NoopErrorHandler),
        }
    }
}

impl CompileOption {
    pub fn scanning(&self) -> ScanOption {
        ScanOption {
            delimiters: self.delimiters.clone(),
            get_text_mode: self.get_text_mode,
        }
    }
    pub fn parsing(&self) -> ParseOption {
        ParseOption {
            whitespace: self.whitespace.clone(),
            preserve_comment: self.preserve_comments.unwrap_or(self.is_dev),
            get_namespace: self.get_namespace,
            get_text_mode: self.get_text_mode,
            is_native_element: self.is_native_tag,
            is_void_tag: self.is_void_tag,
            is_pre_tag: self.is_pre_tag,
            get_builtin_component: self.get_builtin_component,
            is_custom_element: self.is_custom_element,
        }
    }
    pub fn converting(&self) -> ConvertOption {
        ConvertOption {
            get_builtin_component: self.get_builtin_component,
            is_dev: self.is_dev,
            directive_converters: self.directive_converters.clone(),
            need_reactivity: self.need_reactivity,
        }
    }
    pub fn transforming(&self) -> TransformOption {
        let prefix = match self.mode {
            ScriptMode::Function {
                prefix_identifier, ..
            } => prefix_identifier,
            ScriptMode::Module { .. } => true,
        };
        TransformOption {
            prefix_identifier: prefix,
            is_dev: self.is_dev,
        }
    }
    pub fn codegen(&self) -> CodeGenerateOption {
        CodeGenerateOption {
            is_dev: self.is_dev,
            mode: self.mode.clone(),
            source_map: self.source_map,
            decode_entities: self.decode_entities,
        }
    }
}

// TODO: refactor this ownership usage
pub trait TemplateCompiler<'a> {
    type IR;
    type Output;

    fn scan(&self, source: &'a str) -> Tokens<'a>;
    fn parse(&self, tokens: Tokens<'a>) -> AstRoot<'a>;
    fn convert(&self, ast: AstRoot<'a>) -> Self::IR;
    fn transform(&mut self, ir: &mut Self::IR);
    fn generate(&mut self, ir: Self::IR) -> Self::Output;
    fn get_error_handler(&self) -> RcErrHandle;

    fn compile(&mut self, source: &'a str) -> Self::Output {
        let tokens = self.scan(source);
        let ast = self.parse(tokens);
        let mut ir = self.convert(ast);
        self.transform(&mut ir);
        self.generate(ir)
    }
}

pub struct BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    writer: Option<W>,
    passes: Option<P>,
    option: CompileOption,
    scanner: Scanner,
    parser: Parser,
    pd: std::marker::PhantomData<&'a ()>,
}

impl<'a, P, W> BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    pub fn new(writer: W, passes: P, option: CompileOption) -> Self {
        Self {
            writer: Some(writer),
            passes: Some(passes),
            scanner: Scanner::new(option.scanning()),
            parser: Parser::new(option.parsing()),
            option,
            pd: std::marker::PhantomData,
        }
    }
    fn get_converter(&self) -> BaseConverter<'a> {
        let eh = self.get_error_handler();
        let option = self.option.converting();
        BaseConverter {
            err_handle: eh,
            sfc_info: Default::default(),
            option,
        }
    }
    fn get_transformer(&mut self, pass: P) -> BaseTransformer<'a, P> {
        BaseTransformer::new(pass)
    }
    fn get_code_generator(&mut self) -> CodeWriter<'a, W> {
        let option = self.option.codegen();
        let writer = self.writer.take().unwrap();
        CodeWriter::new(writer, option, Default::default())
    }
}

impl<'a, P, W> TemplateCompiler<'a> for BaseCompiler<'a, P, W>
where
    W: io::Write,
    P: CorePass<BaseInfo<'a>>,
{
    type IR = BaseRoot<'a>;
    type Output = io::Result<()>;

    fn scan(&self, source: &'a str) -> Tokens<'a> {
        self.scanner.scan(source, self.get_error_handler())
    }

    fn parse(&self, tokens: Tokens<'a>) -> AstRoot<'a> {
        self.parser.parse(tokens, self.get_error_handler())
    }
    fn convert(&self, ast: AstRoot<'a>) -> Self::IR {
        self.get_converter().convert_ir(ast)
    }
    fn transform(&mut self, ir: &mut Self::IR) {
        let pass = self.passes.take().unwrap();
        self.get_transformer(pass).transform(ir);
    }
    fn generate(&mut self, ir: Self::IR) -> Self::Output {
        self.get_code_generator().generate(ir)
    }
    fn get_error_handler(&self) -> RcErrHandle {
        self.option.error_handler.clone()
    }
}
