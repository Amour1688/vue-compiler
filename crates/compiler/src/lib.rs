#![allow(dead_code, unused_variables)]
//! See README.md

use std::ops::Range;

// TODO: reorg pub
pub mod codegen;
pub mod compiler;
pub mod converter;
pub mod error;
pub mod flags;
pub mod parser;
pub mod scanner;
pub mod transformer;
#[macro_use]
pub mod util;

#[cfg(feature = "serde")]
use serde::Serialize;

// use plain &str here for now
// may change to tendril
pub type Name<'a> = &'a str;

#[derive(PartialEq, Eq, Clone)]
pub struct Position {
    /// the 0-indexed offset in the source str modulo newline
    pub offset: usize,
    /// the line number in the source code
    pub line: u32,
    /// the column number in the source code
    pub column: u32,
}

#[cfg(feature = "serde")]
impl Serialize for Position {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let s = format!(
            // Position, Line, Column
            "Pos: {}, Ln: {}, Col: {}",
            self.offset,
            self.line,
            self.column,
        );
        serializer.serialize_str(&s)
    }
}

impl Default for Position {
    fn default() -> Self {
        Self {
            offset: 0,
            line: 1,
            column: 1,
        }
    }
}

#[derive(Default, PartialEq, Eq, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

impl From<SourceLocation> for Range<usize> {
    fn from(location: SourceLocation) -> Self {
        location.start.offset..location.end.offset
    }
}

/// namespace for HTML/SVG/MathML tag
#[non_exhaustive]
#[derive(Eq, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize))]
pub enum Namespace {
    Html,
    Svg,
    MathMl,
    UserDefined(&'static str),
}

#[cfg(test)]
#[macro_export]
macro_rules! cast {
    ($target: expr, $pat: path) => {{
        if let $pat(a, ..) = $target {
            a
        } else {
            panic!("mismatch variant when cast to {}", stringify!($pat));
        }
    }};
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test_source_size() {
        assert_eq!(std::mem::size_of::<Position>(), 16);
    }
}
