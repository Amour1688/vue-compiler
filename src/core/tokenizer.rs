use std::str::Chars;
use super::{
    Name, SourceLocation, Position,
    error::{CompilationError, CompilationErrorKind},
};

#[derive(PartialEq, Eq, Debug)]
pub enum Token<'a> {
    // bracket related token
    LeftBracket,
    RightBracket,
    LeftBracketSlash,
    SlashRightBracket,

    // in tag token
    TagName(Name<'a>),
    AttrName(Name<'a>),
    Equal,
    Value(Name<'a>),

    // content in raw text
    Text(&'a str),
    Interpolation(&'a str),
    Comment(&'a str),
}

// Note: TokenizerOption is not thread safe
// due to cached_first_char is shared
pub struct TokenizerOption {
    pub delimiters: (String, String),
    // for search optimization: only compare delimiters' first char
    cached_first_char: Option<char>,
}

impl TokenizerOption {
    fn delimiter_first_char(&mut self) -> char {
        if let Some(c) =  self.cached_first_char {
            return c
        }
        let c = self.delimiters.0.chars().next()
            .expect("interpolation delimiter cannot be empty");
        self.cached_first_char.replace(c);
        c
    }
}

impl Default for TokenizerOption {
    fn default() -> Self {
        Self {
            delimiters: ("{{".into(), "}}".into()),
            cached_first_char: Some('{'),
        }
    }
}

#[derive(PartialEq, Eq)]
enum TokenizerState {
    /// match text between <tag> and </tag>
    RawText,
    /// matched left bracket <, expect TagName
    OpenTag,
    /// matched <tag, expect attr or right bracket >
    InTag,
    /// matched attribute name, expect = or right bracket >
    InAttr,
    /// matched attr=, expect attribute value
    AttrEqual,
}

pub struct Tokenizer<'a> {
    source: &'a str,
    position: Position,
    state: TokenizerState,
    option: TokenizerOption,
    errors: Vec<CompilationError>,
}

// builder methods
impl<'a> Tokenizer<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            position: Default::default(),
            state: TokenizerState::RawText,
            option: Default::default(),
            errors: Vec::new(),
        }
    }
    pub fn with_option<'b>(&'b mut self, option: TokenizerOption) -> &'b mut Tokenizer<'a> {
        self.option = option;
        self
    }
}

// scanning methods
impl<'a> Tokenizer<'a> {
    fn scan_rawtext(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::RawText);
        let d = self.option.delimiter_first_char();
        let index = self.source
            .find(|c| c == '<' || c == d);
        // no tag or interpolation found
        if index.is_none() {
            let src = self.move_by(self.source.len());
            return Token::Text(src)
        }
        let i = index.unwrap();
        if i != 0 {
            let src = self.move_by(i);
            return Token::Text(src)
        }
        let source = self.source;
        if source.starts_with(&self.option.delimiters.0) {
            return self.scan_interpolation()
        }
        if source.starts_with("</") {
            return Token::LeftBracketSlash
        }
        if source.starts_with("<!") {
            return self.scan_comment_and_like()
        }
        self.state = TokenizerState::OpenTag;
        self.move_by(1);
        Token::LeftBracket
    }

    fn scan_open_tag(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::OpenTag);
        let chars = self.source.chars();
        let l = scan_tag_name_length(chars);
        if l == 0 {
            self.emit_error(CompilationErrorKind::InvalidFirstCharacterOfTagName);
            self.state = TokenizerState::RawText;
            return self.scan_rawtext()
        }
        self.state = TokenizerState::InTag;
        let src = self.move_by(l);
        Token::TagName(src)
    }

    fn scan_in_tag(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::InTag);
        self.skip_whitespace();
        if self.source.starts_with("/>") {
            return Token::SlashRightBracket
        }
        if self.source.starts_with('>') {
            return Token::RightBracket
        }
        self.scan_attr_name()
    }
    fn scan_attr_name(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::InTag);
        debug_assert!(self.source.starts_with(is_valid_name_char));
        // state must transit to InAttr even if any error occurs
        self.state = TokenizerState::InAttr;
        if self.source.starts_with('=') {
            self.emit_error(CompilationErrorKind::MissingAttributeName);
            return Token::AttrName("")
        }
        let count = self.source.chars()
            .take_while(|&c| semi_valid_attr_name(c))
            .count();
        let src = self.move_by(count);
        if src.contains(|c| matches!(c, '<' | '"' | '\'')) {
            self.emit_error(CompilationErrorKind::UnexpectedCharacterInAttributeName);
        }
        Token::AttrName(src)
    }
    fn scan_in_attr(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::InAttr);
        todo!()
    }
    fn scan_attr_equal(&mut self) -> Token<'a> {
        debug_assert!(self.state == TokenizerState::AttrEqual);
        todo!()
    }

    fn scan_interpolation(&mut self) -> Token<'a> {
        let delimiters = &self.option.delimiters;
        debug_assert!(self.source.starts_with(&delimiters.0));
        let index =  self.source.find(&delimiters.1);
        if index.is_none() {
            self.emit_error(CompilationErrorKind::MissingInterpolationEnd);
            let src = self.move_by(self.source.len());
            return Token::Interpolation(src)
        }
        let step = index.unwrap() + delimiters.1.len();
        let src = self.move_by(step);
        Token::Interpolation(src)
    }

    fn scan_comment_and_like(&mut self) -> Token<'a> {
        let s = self.source;
        if s.starts_with("<!--") {
            self.scan_comment()
        } else if s.starts_with("<!DOCTYPE") {
            self.scan_bogus_comment()
        } else if s.starts_with("<!CDATA[") {
            self.scan_cdata()
        } else {
            self.emit_error(CompilationErrorKind::IncorrectlyOpenedComment);
            self.scan_bogus_comment()
        }
    }
    fn scan_comment(&mut self) -> Token<'a> {
        debug_assert!(self.source.starts_with("<!--"));
        while let Some(end) = self.source.find("--") {
            let s = &self.source[end..];
            let offset = if s.starts_with("!>") {
                2
            } else if s.starts_with('>') {
                1
            } else {
                0
            };
            if offset > 0 {
                let src = self.move_by(end + offset + 2);
                return Token::Comment(src)
            }
        }
        self.emit_error(CompilationErrorKind::EofInComment);
        Token::Comment(self.source)
    }
    #[cold]
    fn scan_bogus_comment(&mut self) -> Token<'a> {
        todo!()
    }
    #[cold]
    fn scan_cdata(&mut self) -> Token<'a> {
        todo!()
    }
}

// util mtehods
impl<'a> Tokenizer<'a> {
    fn emit_error(&mut self, error_kind: CompilationErrorKind) {
        let loc = self.current_location();
        let err = CompilationError::new(error_kind).with_location(loc);
        self.errors.push(err);
    }

    fn current_location(&self) -> SourceLocation {
        todo!()
    }

    /// move tokenizer's interal position forward and return the range of movement
    /// tokenizer's line/column are also updated in the method
    /// note it only moves forward, not backward
    /// `advance_to` is a better name but it collides with iter
    fn move_by(&mut self, size: usize) -> &'a str {
        debug_assert!(size > 0, "tokenizer must move forward");
        let mut lines = 0;
        let mut last_new_line_pos = -1;
        for (i, c) in self.source[..size].chars().enumerate() {
            if c == '\n' {
                lines += 1;
                last_new_line_pos = i as i32;
            }
        }
        let old_source = self.source;
        self.source = &self.source[size..];
        let pos = &mut self.position;
        pos.offset += size;
        pos.line += lines;
        pos.column = if last_new_line_pos == -1 {
            pos.column + size
        } else {
            size - last_new_line_pos as usize
        };
        &old_source[..size]
    }

    fn skip_whitespace(&mut self) {
        let idx = self.source.find(|c: char| !c.is_ascii_whitespace());
        if let Some(i) = idx {
            self.move_by(i);
        } else {
            self.move_by(self.source.len());
        }
    }
}

// `< ' "` are not valid but counted as semi valid
// to leniently recover from a parsing error
#[inline]
fn semi_valid_attr_name(c: char) -> bool {
    is_valid_name_char(c) && c != '='
}

#[inline]
fn is_valid_name_char(c: char) -> bool {
    !c.is_ascii_whitespace() && c != '/' && c != '>'
}

// tag name should begin with [a-zA-Z]
// followed by chars except whitespace, / or >
fn scan_tag_name_length(mut chars: Chars<'_>) -> usize {
    let first_char = chars.next();
    debug_assert!(first_char.is_some());
    if !first_char.unwrap().is_ascii_alphabetic() {
        return 0
    }
    let l = chars
        .take_while(|&c| is_valid_name_char(c))
        .count();
    l + 1
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        if self.source.is_empty() {
            return None
        }
        use TokenizerState as S;
        Some(match self.state {
            S::OpenTag => self.scan_open_tag(),
            S::RawText => self.scan_rawtext(),
            S::InTag => self.scan_in_tag(),
            S::InAttr => self.scan_in_attr(),
            S::AttrEqual => self.scan_attr_equal(),
        })
    }
}

#[cfg(test)]
mod test {
    fn test() {
        let cases = [
            r#"<a v-bind:['foo' + bar]="value">...</a>"#,
        ];
    }
}
