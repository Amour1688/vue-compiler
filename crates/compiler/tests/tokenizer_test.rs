mod common;

use common::{serialize_yaml, SourceLocation, TestErrorHandler};
use compiler::tokenizer::{self, TextMode, TokenizeOption, Tokenizer};
use insta::{assert_snapshot, assert_yaml_snapshot};
use serde::Serialize;

#[derive(Serialize)]
pub struct Attribute {
    pub name: String,
    pub value: Option<AttributeValue>,
    pub name_loc: SourceLocation,
    pub location: SourceLocation,
}

#[derive(Serialize)]
pub struct AttributeValue {
    pub content: String,
    pub location: SourceLocation,
}

#[derive(Serialize)]
pub struct Tag {
    pub name: String,
    pub attributes: Vec<Attribute>,
    pub self_closing: bool,
}

#[derive(Serialize)]
pub enum Token {
    StartTag(Tag),
    EndTag(String),
    Text(String),
    Comment(String),
    Interpolation(String),
}

impl<'a> From<tokenizer::Token<'a>> for Token {
    fn from(t: tokenizer::Token<'a>) -> Self {
        use tokenizer::Token as T;
        match t {
            T::StartTag(s) => Self::StartTag(s.into()),
            T::EndTag(e) => Self::EndTag(e.into()),
            T::Text(t) => Self::Text(t.into_string()),
            T::Comment(c) => Self::Comment(c.into()),
            T::Interpolation(i) => Self::Interpolation(i.into()),
        }
    }
}

impl<'a> From<tokenizer::Tag<'a>> for Tag {
    fn from(t: tokenizer::Tag<'a>) -> Self {
        let attrs = t.attributes.into_iter().map(|a| Attribute {
            name: a.name.into(),
            value: a.value.map(|v| AttributeValue {
                content: v.content.into_string(),
                location: v.location.into(),
            }),
            name_loc: a.name_loc.into(),
            location: a.location.into(),
        });
        Tag {
            name: t.name.into(),
            attributes: attrs.collect(),
            self_closing: t.self_closing,
        }
    }
}

fn scan_with_opt(s: &str, opt: TokenizeOption) -> impl Iterator<Item = Token> + '_ {
    let tokenizer = Tokenizer::new(opt);
    let ctx = TestErrorHandler;
    tokenizer.scan(s, ctx).map(Token::from)
}

pub fn base_scan(s: &str) -> impl Iterator<Item = Token> + '_ {
    scan_with_opt(s, TokenizeOption::default())
}

fn assert_yaml<S: Serialize>(val: S, expr: &str) {
    let name = insta::_macro_support::AutoName;
    let val = serialize_yaml(val);
    assert_snapshot!(name, val, expr);
}

#[test]
fn should_scan() {
    let cases = [
        r#"<![CDATA["#,
        r#"{{}}"#,
        r#"{{test}}"#,
        r#"<a test="value">...</a>"#,
        r#"<a v-bind:['foo' + bar]="value">...</a>"#,
        r#"<tag =value />"#,
        r#"<a =123 />"#,
        r#"<a ==123 />"#,
        r#"<a b="" />"#,
        r#"<a == />"#,
        r#"<a wrong-attr>=123 />"#,
        r#"<a></a < / attr attr=">" >"#,
        r#"<a attr="1123"#,              // unclosed quote
        r#"<a attr=""#,                  // unclosed without val
        r#"<!-->"#,                      // abrupt closing
        r#"<!--->"#,                     // abrupt closing
        r#"<!---->"#,                    // ok
        r#"<!-- nested <!--> text -->"#, // ok
        r#"<p v-err=232/>"#,
    ];
    for case in cases {
        let val: Vec<_> = base_scan(case).collect();
        assert_yaml(val, case);
    }
}

#[test]
fn should_scan_raw_text() {
    let cases = [
        r#"<style></style"#,
        r#"<style></styl"#,
        r#"<style></styles"#,
        r#"<style></style "#,
        r#"<style></style>"#,
        r#"<style>abc</style>"#,
    ];
    for &case in cases.iter() {
        let opt = TokenizeOption {
            get_text_mode: |_| TextMode::RawText,
            ..Default::default()
        };
        let t: Vec<_> = scan_with_opt(case, opt).collect();
        assert_yaml(t, case);
    }
}
#[test]
fn should_scan_rc_data() {
    let cases = [
        r#"<textarea>   "#,
        r#"<textarea></textarea "#,
        r#"<textarea></textarea"#,
        r#"<textarea></textareas>"#,
        r#"<textarea><div/></textarea>"#,
        r#"<textarea><div/></textareas>"#,
        r#"<textarea>{{test}}</textarea>"#,
        r#"<textarea>{{'</textarea>'}}</textarea>"#,
        r#"<textarea>{{}}</textarea>"#,
        r#"<textarea>{{</textarea>"#,
        r#"<textarea>{{"#,
        r#"<textarea>{{ garbage  {{ }}</textarea>"#,
    ];
    for &case in cases.iter() {
        let opt = TokenizeOption {
            get_text_mode: |_| TextMode::RcData,
            ..Default::default()
        };
        let a: Vec<_> = scan_with_opt(case, opt).collect();
        assert_yaml(a, case);
    }
}
