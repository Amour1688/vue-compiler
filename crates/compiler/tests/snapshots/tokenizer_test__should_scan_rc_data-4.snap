---
source: crates/compiler/tests/tokenizer_test.rs
expression: "<textarea></textareas>"

---
- StartTag:
    name: textarea
    attributes: []
    self_closing: false
- Text: "</textareas>"

